### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by Shawdooow**
@Shawdooow  previously said: 
>Don't merge this or I will find you.
>

### CODE REVIEW

File: Lovewing.Game/Graphics/Background.cs [hunk 1]
```diff
@@ -14,31 +14,28 @@ public class Background : BufferedContainer
     {
         public Sprite Sprite;
 
-        private readonly string textureName;
-
-        public Background(string textureName)
+        public Background()
```
Comment (jorolf) [inline Lovewing.Game/Graphics/Background.cs]: you made this class pointless

Comment (Shawdooow) [reply Lovewing.Game/Graphics/Background.cs]: did i?

Comment (jorolf) [reply Lovewing.Game/Graphics/Background.cs]: this is just a container with a sprite.
before it could actually load backgrounds

Comment (Shawdooow) [reply Lovewing.Game/Graphics/Background.cs]: i will fix that then

File: Lovewing.Game/Graphics/Background.cs [hunk 2]
```diff
@@ -13,32 +13,32 @@ namespace Lovewing.Game.Graphics
...
         [BackgroundDependencyLoader]
         private void load(TextureStore textures)
         {
-            if (!string.IsNullOrEmpty(textureName))
-                Sprite.Texture = textures.Get(textureName);
+            Children = new Drawable[]
+            {
+                Sprite = new Sprite
+                {
+                    Texture = textures.Get("Backgrounds\\"+texture),
```
Comment (jorolf) [inline Lovewing.Game/Graphics/Background.cs]: "/" not "\\"

File: Lovewing.Game/Graphics/UserInterface/WedgeButton.cs [hunk 3]
```diff
@@ -12,7 +12,8 @@ namespace Lovewing.Game.Graphics.UserInterface
     public class WedgeButton : Container
     {
         public Color4 ActiveColor;
-        public string Text;
+        public string Text { get; set; } = "null";
```
Comment (jorolf) [inline Lovewing.Game/Graphics/UserInterface/WedgeButton.cs]: did you test this? 99% sure this doesn't work

File: Lovewing.Game/Graphics/UserInterface/WedgeButton.cs [hunk 4]
```diff
@@ -12,7 +12,8 @@ namespace Lovewing.Game.Graphics.UserInterface
...
+        public float TextSize { get; set; } = 20;
```
Comment (jorolf) [inline Lovewing.Game/Graphics/UserInterface/WedgeButton.cs]: same with this

File: Lovewing.Game/Graphics/UserInterface/WedgeButton.cs [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+                                Anchor = Anchor.Centre,
+                                Origin = Anchor.Centre,
+                            }
+                    }
+                }
+            };
+            buttonText.Hide();
+        }
+
+        protected override void LoadComplete()
```
Comment (jorolf) [inline Lovewing.Game/Graphics/UserInterface/WedgeButton.cs]: you can remove this now

File: Lovewing.Game/Graphics/UserInterface/LovewingRectangleButton.cs [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+using System.Collections.Generic;
+using OpenTK;
+using osu.Framework.Graphics.Sprites;
+
+namespace Lovewing.Game.Graphics.UserInterface
+{
+    public class LovewingRectangleButton : BaseLovewingButton, IFilterable
+    {
+        public Color4 LightColor { get; set; }
+        public Color4 DarkColor;
```
Comment (jorolf) [inline Lovewing.Game/Graphics/UserInterface/LovewingRectangleButton.cs]: `LightColor` is a property but `DarkColor` not??

File: Lovewing.Game/Graphics/UserInterface/LovewingRectangleButton.cs [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+        private Container hover;
+        private Circle ripple;
+        private Container button;
+
+        public LovewingRectangleButton()
+        {
+            SpriteText.Shadow = true;
+        }
+
+        protected override void LoadComplete()
```
Comment (jorolf) [inline Lovewing.Game/Graphics/UserInterface/LovewingRectangleButton.cs]: don't use `LoadComplete` to set properties, do that in the setter and getter of the property, that's why they exist

File: Lovewing.Game/Graphics/UserInterface/LovewingRectangleButton.cs [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+        {
+            base.LoadComplete();
+
+            DarkColor = LightColor.Darken(0.2f);
```
Comment (jorolf) [inline Lovewing.Game/Graphics/UserInterface/LovewingRectangleButton.cs]: oh, that's why....
make it a property with only a getter that returns exactly that

File: Lovewing.Game/LovewingTests.cs [hunk 9]
```diff
@@ -22,7 +22,7 @@ protected override void LoadComplete()
                 new CursorContainer(),
             };
 
-            LoadComponentAsync(new Background("Backgrounds/mainmenu") { Depth = 10 }, AddInternal);
+            //LoadComponentAsync(new Background("Backgrounds/mainmenu") { Depth = 10 }, AddInternal);
```
Comment (jorolf) [inline Lovewing.Game/LovewingTests.cs]: why did you comment this out?

File: Lovewing.Game/Screens/LovewingScreen.cs [hunk 10]
```diff
@@ -10,6 +12,12 @@ namespace Lovewing.Game.Screens
     /// </summary>
     public class LovewingScreen : Screen
     {
+        public Background Background = new Background("mainmenu");
 
+        protected override void LoadComplete()
+        {
+            base.LoadComplete();
+            AddRange(new Drawable[] { Background });
```
Comment (jorolf) [inline Lovewing.Game/Screens/LovewingScreen.cs]: why in `LoadComplete`?

File: Lovewing.Game/Screens/Main/Wedge.cs [hunk 11]
```diff
@@ -85,26 +83,28 @@ public override bool Invalidate(Invalidation invalidation = Invalidation.All, Dr
...
 
         protected override void PopOut()
         {
-            wedgeBackground.MoveTo(new Vector2(2, 0), 250, Easing.InQuad);
-            content.MoveTo(new Vector2(content.DrawSize.X, 0), 250, Easing.InQuad);
+            this.MoveTo(new Vector2(content.DrawSize.X + 100, 0), 250, Easing.InQuad);
         }
 
-        public Drawable CreateButton(string text = "")
+        public WedgeButton CreateButton(string text, Vector2 offset)
```
Comment (jorolf) [inline Lovewing.Game/Screens/Main/Wedge.cs]: you probably shouldn't need this

File: Lovewing.Game/Screens/MainScreen.cs [hunk 12]
```diff
@@ -72,171 +73,148 @@ public MainScreen()
...
             liveshow.StateChanged += vis => selectWedge(liveshow, vis);
 
             AddRange(new[]
             {
-                liveshow.CreateButton("Liveshow"),
-                management.CreateButton("Idols"),
-                home.CreateButton("Home")
+                liveshow.CreateButton("Liveshow", new Vector2(-50, 0)),
+                management.CreateButton("Idols", new Vector2(50, 0)),
+                home.CreateButton("Home", new Vector2(0, 0))
```
Comment (jorolf) [inline Lovewing.Game/Screens/MainScreen.cs]: if you're going to do this use the empty constructor

File: Lovewing.Game/Tests/Visual/TestCaseWedge.cs [hunk 13]
```diff
@@ -50,7 +50,7 @@ public TestCaseWedge()
                 });
 
                 Add(wedge);
-                Add(wedge.CreateButton());
+                Add(wedge.CreateButton("test", Vector2.Zero));
```
Comment (jorolf) [inline Lovewing.Game/Tests/Visual/TestCaseWedge.cs]: or this

File: Lovewing.Game/Graphics/UserInterface/LovewingCircleButton.cs [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+                            Alpha = 0,
+                            Children = new Drawable[]
+                            {
+                                new Box
+                                {
+                                    RelativeSizeAxes = Axes.Both
+                                }
+                            }
+                        },
+                        ripple = new Circle
```
[the remark this answers is not in the document]
Comment (Shawdooow) [reply Lovewing.Game/Graphics/UserInterface/LovewingCircleButton.cs]: you told me to put a circle here, do you want it or not?

File: Lovewing.Game/Graphics/UserInterface/LovewingRectangleButton.cs (hunk 7 above)
Comment (Shawdooow) [reply Lovewing.Game/Graphics/UserInterface/LovewingRectangleButton.cs]: cant here, look again

File: Lovewing.Game/LovewingTests.cs (hunk 9 above)
Comment (Shawdooow) [reply Lovewing.Game/LovewingTests.cs]: its was broken while i was testing, will fix before final pr

File: Lovewing.Game/Screens/LovewingScreen.cs (hunk 10 above)
Comment (Shawdooow) [reply Lovewing.Game/Screens/LovewingScreen.cs]: because it is broken as fuck and i tried to fix it

File: Lovewing.Game/Screens/Main/Wedge.cs (hunk 11 above)
Comment (Shawdooow) [reply Lovewing.Game/Screens/Main/Wedge.cs]: look at why its used

File: Lovewing.Game/Graphics/UserInterface/LovewingCircleButton.cs (hunk 14 above)
[the remark this answers is not in the document]
Comment (jorolf) [reply Lovewing.Game/Graphics/UserInterface/LovewingCircleButton.cs]: what I mean is that circle already provides a box so you don't have to overwrite it

File: Lovewing.Game/Graphics/UserInterface/LovewingRectangleButton.cs [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+                                    Origin = Anchor.CentreRight,
+                                    Width = Size.X * 0.8f,
+                                    EdgeSmoothness = Vector2.One,
+                                    Shear = new Vector2(-0.4f, 0),
+                                    Depth = -1,
+                                    RelativeSizeAxes = Axes.Y,
+                                },
+                                ripple = new Circle
+                                {
+                                    Depth = -2,
```
Comment (jorolf) [inline Lovewing.Game/Graphics/UserInterface/LovewingRectangleButton.cs]: instead of using depth, can't you just reorder them here?

File: Lovewing.Game/Graphics/UserInterface/LovewingCircleButton.cs (hunk 14 above)
[the remark this answers is not in the document]
Comment (Shawdooow) [reply Lovewing.Game/Graphics/UserInterface/LovewingCircleButton.cs]: ._.

File: Lovewing.Game/Screens/Main/Wedge.cs (hunk 11 above)
Comment (jorolf) [reply Lovewing.Game/Screens/Main/Wedge.cs]: I did, you can just do that there

File: Lovewing.Game/Graphics/UserInterface/LovewingRectangleButton.cs (hunk 15 above)
Comment (Shawdooow) [reply Lovewing.Game/Graphics/UserInterface/LovewingRectangleButton.cs]: Does that work now? I remember in vitaru it was alphabetical last time i tried

File: Lovewing.Game/Screens/MainScreen.cs (hunk 12 above)
Comment (jorolf) [reply Lovewing.Game/Screens/MainScreen.cs]: `new Vector2()`

Comment (Shawdooow) [reply Lovewing.Game/Screens/MainScreen.cs]: 👍

File: Lovewing.Game/Screens/LovewingScreen.cs [hunk 16]
```diff
@@ -12,7 +12,7 @@ namespace Lovewing.Game.Screens
     /// </summary>
     public class LovewingScreen : Screen
     {
-        public Background Background = new Background("mainmenu");
+        public Background Background { get; set; }
```
Comment (jorolf) [inline Lovewing.Game/Screens/LovewingScreen.cs]: why not just do the whole logic in specific screens/ why did you change this?

File: Lovewing.Game/Graphics/UserInterface/BaseLovewingButton.cs [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+namespace Lovewing.Game.Graphics.UserInterface
+{
+    public class BaseLovewingButton : ClickableContainer
+    {
+        public string Text
+        {
+            get { return SpriteText?.Text; }
+            set
+            {
+                if (SpriteText != null)
```
[the remark this answers is not in the document]
Comment (Shawdooow) [reply Lovewing.Game/Graphics/UserInterface/BaseLovewingButton.cs]: this was copied from `Button`

File: Lovewing.Game/Screens/LovewingScreen.cs (hunk 16 above)
Comment (Shawdooow) [reply Lovewing.Game/Screens/LovewingScreen.cs]: This is my poor attempt to fix the bg, which failed again >:C

File: Lovewing.Game/Graphics/UserInterface/LovewingRectangleButton.cs (hunk 15 above)
Comment (jorolf) [reply Lovewing.Game/Graphics/UserInterface/LovewingRectangleButton.cs]: "alphabetical"

File: Lovewing.Game/Graphics/UserInterface/LovewingRectangleButton.cs (hunk 6 above)
Comment (binsenpai) [reply Lovewing.Game/Graphics/UserInterface/LovewingRectangleButton.cs]: same

File: Lovewing.Game/Graphics/UserInterface/BaseLovewingButton.cs (hunk 17 above)
[the remark this answers is not in the document]
Comment (sr229) [reply Lovewing.Game/Graphics/UserInterface/BaseLovewingButton.cs]: Most cases SpriteText is always filled, just return a NullRef if null

File: Lovewing.Game/Graphics/UserInterface/LovewingCircleButton.cs [hunk 18]
```diff
@@ -0,0 +1,118 @@
...
+using osu.Framework.Input;
+using OpenTK.Graphics;
+using System.Collections.Generic;
+using OpenTK;
+
+namespace Lovewing.Game.Graphics.UserInterface
+{
+    public class LovewingCircleButton : BaseLovewingButton, IFilterable
+    {
+        private readonly CircularContainer hover;
```
Comment (sr229) [inline Lovewing.Game/Graphics/UserInterface/LovewingCircleButton.cs]: I don't get why not a regular Container instead of Circular for hover

File: Lovewing.Game/Graphics/UserInterface/BaseLovewingButton.cs (hunk 17 above)
[the remark this answers is not in the document]
Comment (Shawdooow) [reply Lovewing.Game/Graphics/UserInterface/BaseLovewingButton.cs]: this is copied straight from the framework, i suggest you fix this upstream if its a point of concern

File: Lovewing.Game/Graphics/UserInterface/LovewingRectangleButton.cs [hunk 19]
```diff
@@ -0,0 +1,149 @@
...
+                    RelativeSizeAxes = Axes.Both,
+
+                    Children = new Drawable[]
+                    {
+                        new Container
+                        {
+                            Anchor = Anchor.Centre,
+                            Origin = Anchor.Centre,
+                            CornerRadius = ButtonCornerRadius,
+                            Depth = 0,
```
Comment (sr229) [inline Lovewing.Game/Graphics/UserInterface/LovewingRectangleButton.cs]: not sure what Depth does but if it's regarding shadow you probably you need to make it Depth = 1

Comment (Shawdooow) [reply Lovewing.Game/Graphics/UserInterface/LovewingRectangleButton.cs]: nope, what is rendered on top of what

File: Lovewing.Game/Screens/MainScreen.cs [hunk 20]
```diff
@@ -25,6 +25,9 @@ public class MainScreen : LovewingScreen
         private readonly LovewingColors colors = new LovewingColors();
         private readonly Wedge liveshow;
 
+        private const float home_button_height = 200;
+        private const float x_offset = -100;
+
         private void solo()
         {
             liveshow.Expand();
```
Comment (sr229) [inline Lovewing.Game/Screens/MainScreen.cs]: Should also hide the buttons and Pop in the back button from MatchmakingScreen

File: Lovewing.Game/Screens/LovewingScreen.cs [hunk 21]
```diff
@@ -10,6 +12,12 @@ namespace Lovewing.Game.Screens
     /// </summary>
```
Comment (sr229) [inline Lovewing.Game/Screens/LovewingScreen.cs]: why is there a stray XMLDoc here?

File: Lovewing.Game/Screens/MainScreen.cs [hunk 22]
```diff
@@ -38,17 +41,16 @@ private void matchmaking()
         public MainScreen()
         {
             // Wedge home, management, liveshow;
```
Comment (sr229) [inline Lovewing.Game/Screens/MainScreen.cs]: L43 should be removed too tbh

File: Lovewing.Game/Screens/LovewingScreen.cs (hunk 21 above)
Comment (Shawdooow) [reply Lovewing.Game/Screens/LovewingScreen.cs]: it was there before i think?

**Comment by Shawdooow** [comment]
This pr should be ready to go now imo

**Comment by sr229** [comment]
waiting for @jorolf

File: Lovewing.Game/Graphics/UserInterface/LovewingRectangleButton.cs [hunk 23]
```diff
@@ -0,0 +1,149 @@
...
+
+        protected override bool OnMouseUp(InputState state, MouseUpEventArgs args)
+        {
+            Content.ScaleTo(1, 1000, Easing.OutElastic);
+            return base.OnMouseUp(state, args);
+        }
+
+        protected override bool OnClick(InputState state)
+        {
+            ripple.FadeOutFromOne(450, Easing.OutCirc);
```
Comment (noud02) [inline Lovewing.Game/Graphics/UserInterface/LovewingRectangleButton.cs]: ripple should be scaled and appear at the cursor pos, maybe you should look through the [material design guidelines](https://material.io/guidelines/motion/material-motion.html)

File: Lovewing.Game/Graphics/UserInterface/LovewingCircleButton.cs [hunk 24]
```diff
@@ -0,0 +1,118 @@
...
+
+        protected override bool OnMouseUp(InputState state, MouseUpEventArgs args)
+        {
+            Content.ScaleTo(1, 1000, Easing.OutElastic);
+            return base.OnMouseUp(state, args);
+        }
+
+        protected override bool OnClick(InputState state)
+        {
+            ripple.FadeOutFromOne(450, Easing.OutCirc);
```
Comment (noud02) [inline Lovewing.Game/Graphics/UserInterface/LovewingCircleButton.cs]: same here

File: Lovewing.Game/Graphics/Background.cs [hunk 25]
```diff
@@ -1,44 +1,45 @@
...
         [BackgroundDependencyLoader]
         private void load(TextureStore textures)
         {
-            if (!string.IsNullOrEmpty(textureName))
-                Sprite.Texture = textures.Get(textureName);
+            Children = new Drawable[]
+            {
+                new Sprite
+                {
+                    Texture = textures.Get("Backgrounds/"+texture),
```
Comment (jorolf) [inline Lovewing.Game/Graphics/Background.cs]: set the Texture after `RelativeSizeAxes` and remove the Size = ...

File: Lovewing.Game/Graphics/UserInterface/LovewingCircleButton.cs [hunk 26]
```diff
@@ -0,0 +1,118 @@
...
+                    }
+                },
+            };
+
+            SpriteText.Shadow = true;
+        }
+
+        protected override bool OnHover(InputState state)
+        {
+            hover.FadeTo(0.5f, 200);
```
Comment (jorolf) [inline Lovewing.Game/Graphics/UserInterface/LovewingCircleButton.cs]: what exactly does this do and why can't it be done in another way?

File: Lovewing.Game/Graphics/UserInterface/LovewingCircleButton.cs [hunk 27]
```diff
@@ -0,0 +1,118 @@
...
+using osu.Framework.Graphics.Shapes;
+using osu.Framework.Graphics.Containers;
+using osu.Framework.Input;
+using OpenTK.Graphics;
+using System.Collections.Generic;
+using OpenTK;
+
+namespace Lovewing.Game.Graphics.UserInterface
+{
+    public class LovewingCircleButton : BaseLovewingButton, IFilterable
```
Comment (jorolf) [inline Lovewing.Game/Graphics/UserInterface/LovewingCircleButton.cs]: remove the IFilterable stuff, we don't need it rn

File: Lovewing.Game/Graphics/UserInterface/LovewingRectangleButton.cs [hunk 28]
```diff
@@ -0,0 +1,149 @@
...
+using osu.Framework.Input;
+using osu.Framework.Extensions.Color4Extensions;
+using OpenTK.Graphics;
+using System.Collections.Generic;
+using OpenTK;
+using osu.Framework.Graphics.Sprites;
+
+namespace Lovewing.Game.Graphics.UserInterface
+{
+    public class LovewingRectangleButton : BaseLovewingButton, IFilterable
```
Comment (jorolf) [inline Lovewing.Game/Graphics/UserInterface/LovewingRectangleButton.cs]: same here, remove that
_or put it in the base button if we really need it_

File: Lovewing.Game/Graphics/UserInterface/WedgeButton.cs [hunk 29]
```diff
@@ -0,0 +1,100 @@
...
+            }
+        }
+
+        public Action Action;
+
+        public WedgeButton()
+        {
+            Children = new Drawable[] 
+            {
+                clickCon = new ClickableContainer
```
Comment (jorolf) [inline Lovewing.Game/Graphics/UserInterface/WedgeButton.cs]: why not just make the whole class a `ClickableContainer` instead of placing one inside it?

File: Lovewing.Game/Graphics/UserInterface/WedgeButton.cs [hunk 30]
```diff
@@ -0,0 +1,100 @@
...
+                    Action = () => Action?.Invoke(),
+                    EdgeEffect = new EdgeEffectParameters
+                    {
+                        Colour = Color4.Black.Opacity(0.25f),
+                        Type = EdgeEffectType.Shadow,
+                        Radius = 10
+                    },
+                    Children = new Drawable[]
+                    {
+                            background = new Box
```
Comment (jorolf) [inline Lovewing.Game/Graphics/UserInterface/WedgeButton.cs]: one indent to much

File: Lovewing.Game/Screens/LovewingScreen.cs (hunk 21 above)
Comment (jorolf) [reply Lovewing.Game/Screens/LovewingScreen.cs]: It's not stray.....
I added that so others will know why I created this class

File: Lovewing.Game/Screens/MainScreen.cs [hunk 31]
```diff
@@ -72,171 +74,148 @@ public MainScreen()
...
                         },
                         management = new IdolManagementWedge
                         {
-                            Anchor = Anchor.BottomRight,
-                            Origin = Anchor.BottomRight,
-                            RelativeSizeAxes = Axes.Both,
-                            Width = 0.5f,
-                            Margin = new MarginPadding { Right = 50 },
+                            RelativeSizeAxes = Axes.Y,
+                            Position = new Vector2(108 + 50, 0),
```
Comment (jorolf) [inline Lovewing.Game/Screens/MainScreen.cs]: try to avoid setting position pls

Comment (jorolf) [reply Lovewing.Game/Screens/MainScreen.cs]: try to do stuff with anchor/origin, `RelativeSizeAxes` and `FlowContainer`s

File: Lovewing.Game/Graphics/UserInterface/LovewingCircleButton.cs (hunk 26 above)
Comment (Shawdooow) [reply Lovewing.Game/Graphics/UserInterface/LovewingCircleButton.cs]: it makes the hover circle appear?

File: Lovewing.Game/Graphics/UserInterface/WedgeButton.cs (hunk 29 above)
Comment (Shawdooow) [reply Lovewing.Game/Graphics/UserInterface/WedgeButton.cs]: that made it not function right, this is easier and tbh not that much grosser

File: Lovewing.Game/Graphics/Background.cs (hunk 25 above)
Comment (Shawdooow) [reply Lovewing.Game/Graphics/Background.cs]: isn't it cleaner to just do it all at once?

File: Lovewing.Game/Screens/MainScreen.cs (hunk 31 above)
Comment (Shawdooow) [reply Lovewing.Game/Screens/MainScreen.cs]: `FlowContainer`?

**Comment by Shawdooow** [comment]
Surpassed by #25
